exec/cnex/cnex.c

```diff
@@ -331,8 +337,83 @@ void exec_raiseLiteral(TExecutor *self, TString *name, TString *info, Number cod
     }
     fprintf(stderr, "Unhandled exception %s (%s) (code %d)\n", TCSTR(name), TCSTR(info), number_to_sint32(code));
     cell_freeCell(exceptionvar);
-    // ToDo: Initiate a stack dump here...
+    cJSON *symbols = NULL;
+    cJSON *pStart = NULL;
+    while (self->ip < self->module->bytecode->codelen) {
+        char *dbgpath = strdup(self->module->source_path);
+        dbgpath[strlen(self->module->source_path)-1] = 'd';
+
+        FILE *fp = fopen(dbgpath, "rb");
+        // Don't report out of memory errors, let the stack unwinder report the best it can.
+        if (fp != NULL) {
+            fseek(fp, 0, SEEK_END);
+            long nSize = ftell(fp);
+            fseek(fp, 0, SEEK_SET);
+            char *debug = malloc(nSize);
+            if (debug != NULL) {
+                fread(debug, 1, nSize, fp);
+                fclose(fp);
+                free(dbgpath);
+                symbols = cJSON_Parse(debug);
+                // Remember the start of the symbols, so we can always go back to them.
+                pStart = symbols;
+                free(debug);
+            }
+        }
 
+        if (symbols != NULL) {
+            uint64_t p = self->ip;
+            cJSON *source = cJSON_GetObjectItem(symbols, "source");
+            char hash_string[65];
+            for (int i = 0; i < 32; i++) {
+                snprintf(&hash_string[i*2], 2, "%2.2x", self->module->bytecode->source_hash[i]);
+            }
+            hash_string[64] = '\0';
+            if (strcmp(hash_string, cJSON_GetObjectItem(symbols, "source_hash")->valuestring) != 0) {
+                fprintf(stderr, "  Stack frame #%d: file %s address %d (no debug info available or debug symbols are out of date)\n", self->callstacktop+1, self->module->source_path, self->ip);
+                goto nextframe;
+            }
+            cJSON *jip = NULL;
+            cJSON *jln = NULL;
+            cJSON *lines = NULL;
+            for (;;) {
+                lines = cJSON_GetObjectItem(symbols, "line_numbers");
+                cJSON_ArrayForEach(symbols, lines)
+                {
+                    jip = cJSON_GetArrayItem(symbols, 0);
+                    if ((uint64_t)jip->valueint == p) {
+                        jln = cJSON_GetArrayItem(symbols, 1);
+                        break;
+                    }
+                }
+                if ((uint64_t)jip->valueint == p) {
```

Comment:
Can you just check for `jln != NULL` here instead of repeating the equality test?

Reply:
Fixed.